### PR TITLE
Introduce a Secret/ directory

### PR DESCRIPTION
**Pull request by dadgar**
This PR introduces a `secrets/` directory and a `NOMAD_SECRET_DIR` environment variable

### CODE REVIEW

File: client/allocdir/alloc_dir.go [hunk 1]
```diff
@@ -46,11 +46,23 @@ var (
...
 	TaskDirs = []string{"tmp"}
 )
 
+type CreateSecretDirFn func(allocID, task string) (path string, err error)
+
 type AllocDir struct {
+	// AllocID is the allocation ID for this directory
+	AllocID string
+
+	// TODO
```
Comment (diptanu) [inline client/allocdir/alloc_dir.go]: What is the `TODO` here?

Comment (schmichael) [inline client/allocdir/alloc_dir.go]: TODO?

File: client/allocdir/alloc_dir.go [hunk 2]
```diff
@@ -123,6 +136,10 @@ func NewAllocDir(allocDir string, maxSize int) *AllocDir {
 	return d
 }
 
+func (d *AllocDir) SetSecretDirFn(fn CreateSecretDirFn) {
```
Comment (diptanu) [inline client/allocdir/alloc_dir.go]: Comment

File: client/allocdir/alloc_dir.go [hunk 3]
```diff
@@ -110,8 +122,9 @@ type AllocDirFS interface {
 
 // NewAllocDir initializes the AllocDir struct with allocDir as base path for
 // the allocation directory and maxSize as the maximum allowed size in megabytes.
-func NewAllocDir(allocDir string, maxSize int) *AllocDir {
+func NewAllocDir(allocID, allocDir string, maxSize int) *AllocDir {
```
Comment (diptanu) [inline client/allocdir/alloc_dir.go]: Allow the user to pass the secret dir fn here.

File: client/allocdir/alloc_dir.go [hunk 4]
```diff
@@ -223,6 +252,20 @@ func (d *AllocDir) Build(tasks []*structs.Task) error {
...
+		if d.createSecretDirFn != nil {
+			sdir, err := d.createSecretDirFn(d.AllocID, t.Name)
+			if err != nil {
+				return fmt.Errorf("Creating secret directory for task %q failed: %v", t.Name, err)
+			}
+
+			// Mount the secret directory
+			taskSecret := filepath.Join(taskDir, TaskSecrets)
+			if err := d.mount(sdir, taskSecret); err != nil {
+				return fmt.Errorf("failed to mount secret directory: %v", err)
```
Comment (schmichael) [inline client/allocdir/alloc_dir.go]: Include task name or path?

File: client/secretdir/secret_dir.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package secretdir
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+const ()
```
Comment (schmichael) [inline client/secretdir/secret_dir.go]: remove

File: client/secretdir/secret_dir_unix_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package secretdir
```
Comment (schmichael) [inline client/secretdir/secret_dir_unix_test.go]: Remove file?

**Comment by schmichael** [comment]
LGTM

File: client/secretdir/secret_dir.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+// init checks the secret directory exists and if it doesn't creates the secret
+// directory
+func (s *SecretDir) init() error {
+	if _, err := os.Stat(s.Dir); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to stat secret dir: %v", err)
+	}
+
+	// TODO this shouldn't be hardcoded
+	if err := s.create(32); err != nil {
```
Comment (diptanu) [inline client/secretdir/secret_dir.go]: Extract this in a `const` till we make it configurable.

**Pull request merged**